### PR TITLE
Testable BatchResponse

### LINKED ISSUE #388: BatchResponse implementation makes testing difficult.
* Operating System version: N/A
  * Firebase SDK version: 2
  * Library version: 6.12.2
  * Firebase Product: Messaging

#### Description:

BatchResponse is implemented as a final class with a package-private constructor. As such, this means that as a consumer of the library I have to either wrap the response object or use reflection to make it testable.

#### Steps to reproduce:

```java
// This will not compile
new BatchResponse(responses);
```

#### Suggestions

I would be happy to submit a PR for this but I would like guidance on implementation preference.

a) Remove the final keyword from the BatchResponse class so tools like Mockito can mock the class.
b) Make the constructor public
c) Add a builder or some other method of instantiating the class. This doesn't seem like a useful change.

### PR DESCRIPTION
**Pull request by ToxicBakery**
Resolves #388 
- Converts batch response to an interface
- Moves batch response implementation to BatchResponseImpl
- Update FirebaseMessagingClientImpl to use BatchResponseImpl

### CODE REVIEW

File: src/main/java/com/google/firebase/messaging/BatchResponseImpl.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import com.google.firebase.internal.NonNull;
+
+import java.util.List;
+
+/**
+ * Response from an operation that sends FCM messages to multiple recipients.
+ * See {@link FirebaseMessaging#sendAll(List)} and {@link
+ * FirebaseMessaging#sendMulticast(MulticastMessage)}.
+ */
+public class BatchResponseImpl implements BatchResponse {
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/messaging/BatchResponseImpl.java:29]: Make this package-protected

Comment (ToxicBakery) [reply src/main/java/com/google/firebase/messaging/BatchResponseImpl.java:29]: Good call, apologies

**Comment by hiranya911** [comment]
Hey @ToxicBakery. I just pushed a change to our CI config, so that they will be run on external contributions such as this PR. Can I ask you to please sync against the latest master branch, and make a push?

**Comment by ToxicBakery** [comment]
Done, appears to have run successfully :+1:

**Pull request merged**